text/0000-public-private-dependencies.md

```diff
@@ -0,0 +1,199 @@
+- Feature Name: `public_private_dependencies`
+- Start Date: 2017-04-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Introduce a public/private distinction to crate dependencies.
+
+# Motivation
+[motivation]: #motivation
+
+The crates ecosystem has greatly expanded since Rust 1.0 and with that a few patterns for
+dependencies have evolved that challenge the currently existing dependency declaration
+system in cargo and rust.  The most common problem is that a crate `A` depends on
+another crate `B` but some of the types from crate `B` are exposed through the API in
+crate `A`.  This causes problems in practice if that dependency `B` is also used by the
+user's code itself which often leaves users in less than ideal situations where either
+their code refuses to compile because different versions of those libraries are requested
+or where compiler messages are less than clear.
+
+The introduction of an explicit distinction between public and private dependencies can
+solve some of these issues and also let us lift some restrictions that should make some
+code compile that previously was prevented from compiling by restrictions in cargo.
+
+**Q: What is a public dependency?**<br>
```

Comment:
I don't really like this language. If you depend on items from a dependency (e.g. types) even if you don't reexport them, you still need a public dep. Also----and this is crucial---if you don't reexport anything, it's *not* a breaking change to change your public deps if you're own interface doesn't change.

Consider a situation where `a` depends (privately or publicly) on `b` which publicly depends on `c`. To use any items from `b` with interfaces relying on `c`, `a` should also need to depend on `c` (and constrain that dep to be unified with `b`'s). But now the extra dep of `a` on `c` forces the "effective interface: of `b with given c` to not break. Conversely, if `a` doesn't do that, then we know `b`'s use of `c` is irrelevant so the relevant parts of `b` also won't break.

This is an instance of the general principle that the version of a crate is like a *fallback / catchall* that describes the remnants of the interface not already accounted for in the crate metadata, and constrainable with a dependency.

This may sounds like needleslys fancy reasoning, but this is actually really important practically. Breaking changes already are difficult growing pain with large ecosystems, and would become *impossibly* so if on every upstream breaking change, all downstream was forced to issue their own breaking change even if all they did is bump a dependency.